### PR TITLE
Refactor on blur and on focus field validation actions

### PR DESCRIPTION
**Pull request by ggomez-cambriasolutions**
### Jira Story

Tech debt cleanup - No story

## Description
This PR removes the custom error check actions and reducers for the search UI field validations.

They are replaced with a generalized action and reducer that accepts arguments which set the correct value for the error check.

## Tests
- [x] I have included unit tests 
- [ ] I have included feature tests 
- [ ] I have included other tests 
- [ ] I have NOT included tests 

## Types of changes

- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have ran all tests for the project.
- [x] I have ran lint/rubocop check for the changed/new files.
- [ ] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build, to check Lint/Rubocop, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

File: app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx [hunk 1]
```diff
@@ -8,64 +8,62 @@ import {PersonSearchFieldsPropType, PersonSearchFieldsDefaultProps} from 'data/p
...
+      gridClassName="col-md-3 sex-at-birth-field"
+      onChange={onChange}
+      value={personSearchFields.searchSexAtBirth}
+      onKeyPress={onKeyPress}
+    />
+    <MaskedInputField
+      id="search-client-id"
+      label="Client ID Number"
+      gridClassName="col-md-3 client-id-field"
+      onBlur={() => onBlur('clientIdErrorCheck')}
```
Comment (wmitchell-cambria) [inline app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx]: These anonymous functions will be redefined each render, so they will never compare as equal. This tends to lead to extra React rendering of components whose props did not "actually" change. If possible, pull this out into a method on the class, and reference that directly.

File: app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx [hunk 2]
```diff
@@ -8,64 +8,62 @@ import {PersonSearchFieldsPropType, PersonSearchFieldsDefaultProps} from 'data/p
...
+      onChange={({target: {value}}) => onChange('searchClientId', value)}
+      onFocus={() => onFocus('clientIdErrorCheck')}
```
Comment (wmitchell-cambria) [inline app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx]: here

File: app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx [hunk 3]
```diff
@@ -8,64 +8,62 @@ import {PersonSearchFieldsPropType, PersonSearchFieldsDefaultProps} from 'data/p
...
+      maxLength='19'
+      errors={clientIdError}
+      moveCursor={moveCursor}
+      onKeyPress={onKeyPress}
+    />
+    <MaskedInputField
+      id="search-ssn"
+      label="Social Security Number"
+      gridClassName="col-md-3 ssn-field"
+      onBlur={() => onBlur('ssnErrorCheck')}
```
Comment (wmitchell-cambria) [inline app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx]: here

File: app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx [hunk 4]
```diff
@@ -8,64 +8,62 @@ import {PersonSearchFieldsPropType, PersonSearchFieldsDefaultProps} from 'data/p
...
+      onChange={({target: {value}}) => onChange('searchSsn', value)}
+      onFocus={() => onFocus('ssnErrorCheck')}
```
Comment (wmitchell-cambria) [inline app/javascript/common/search/PersonSearchAgeGenderNumbersGroup.jsx]: here

File: app/javascript/common/search/age/DateOfBirthDateField.jsx [hunk 5]
```diff
@@ -13,7 +13,7 @@ class DateOfBirthDateField extends React.Component {
           gridClassName="date-field"
           label="Date"
           value={value}
-          onBlur={onBlur}
+          onBlur={() => onBlur('dobErrorCheck')}
```
Comment (wmitchell-cambria) [inline app/javascript/common/search/age/DateOfBirthDateField.jsx]: here

**Comment by wmitchell-cambria** [comment]
I'm going to paste the code that CodeClimate is talking about, for ease of discussion.

```javascript
    <MaskedInputField
      id="search-client-id"
      label="Client ID Number"
      gridClassName="col-md-3 client-id-field"
      onBlur={() => onBlur('clientIdErrorCheck')}
      onChange={({target: {value}}) => onChange('searchClientId', value)}
      onFocus={() => onFocus('clientIdErrorCheck')}
      value={personSearchFields.searchClientId}
      mask='1111-1111-1111-1111111'
      placeholder='____-____-____-_______'
      maxLength='19'
      errors={clientIdError}
      moveCursor={moveCursor}
      onKeyPress={onKeyPress}
    />
    <MaskedInputField
      id="search-ssn"
      label="Social Security Number"
      gridClassName="col-md-3 ssn-field"
      onBlur={() => onBlur('ssnErrorCheck')}
      onChange={({target: {value}}) => onChange('searchSsn', value)}
      onFocus={() => onFocus('ssnErrorCheck')}
      value={personSearchFields.searchSsn}
      mask='111-11-1111'
      placeholder='___-__-____'
      maxLength='9'
      errors={ssnErrors}
      moveCursor={moveCursor}
      onKeyPress={onKeyPress}
    />
```

I think there are a couple ways to reduce this duplication.
1. We could create a new component called `HyphenatedMaskedInputField`. That component could take in the `mask` or `placeholder` props above and render a `MaskedInputField` with the appropriate mask, placeholder, maxLength. The rest of the props would be passed through.
2. A lot of the duplication is sooo close to being easily combined. The difficulty is that you have lots of strings that are almost the same: `search-ssn`, `ssn-field`, `searchSsn`, `ssnErrorCheck`. One thing you could do is have a helper method (or new component) which takes in the `fieldName` as a prop (possibly both `fieldName` and `snakeFieldName`), and then generate many of the props based on that.
3. If you have (2) in place, you can combine several of the props that have to be sent to both. Specifically, create an actions object containing `{ onBlur, onFocus, onChange, moveCursor, onKeyPress }` and spread that each.

If you combined the two, you could end up with something like this:

```javascript
<HyphenatedMaskedInputField
  name="client-id"
  label="Client ID Number"
  value={personSearchFields.searchClientId}
  mask='1111-1111-1111-1111111'
  errors={clientIdError}
  {...actions}
/>
<HyphenatedMaskedInputField
  name="ssn"
  label="Social Security Number"
  value={personSearchFields.searchSsn}
  mask='111-11-1111'
  errors={ssnErrors}
  {...actions}
/>
```

File: app/javascript/common/MaskedInputField.jsx [hunk 6]
```diff
@@ -85,13 +86,12 @@ const MaskedInputField = ({
           placeholder={''}
           required={required}
           aria-required={required}
-          onBlur={(event) => {
-            event.target.placeholder = ''
-            if (onBlur) {
-              onBlur(id)
-            }
+          onBlur={handleBlur}
+          onFocus={(e) => {
```
Comment (pokharelmanish) [inline app/javascript/common/MaskedInputField.jsx]: Can we store this arrow function in a `const` and call it here, similar to `handleBlur`.

**Pull request merged**